File: tools/base_persistence_layout.zig
const std = @import("std");
const base = @import("vst3-zig").pluginterfaces.base;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    try stdout.print("FVariant.kEmpty {}\n", .{base.fvariant.FVariant.kEmpty});
    try stdout.print("FVariant.kInteger {}\n", .{base.fvariant.FVariant.kInteger});
    try stdout.print("FVariant.kFloat {}\n", .{base.fvariant.FVariant.kFloat});
    try stdout.print("FVariant.kString8 {}\n", .{base.fvariant.FVariant.kString8});
    try stdout.print("FVariant.kObject {}\n", .{base.fvariant.FVariant.kObject});
    try stdout.print("FVariant.kOwner {}\n", .{base.fvariant.FVariant.kOwner});
    try stdout.print("FVariant.kString16 {}\n", .{base.fvariant.FVariant.kString16});
    try printType(stdout, "FVariant", base.fvariant.FVariant);
    try printOffset(stdout, "FVariant", "type", base.fvariant.FVariant, "type");
    try printOffset(stdout, "FVariant", "intValue", base.fvariant.FVariant, "value");

    try printTuid(stdout, "ICloneable", base.icloneable.icloneable_iid);
    try printTuid(stdout, "IPersistent", base.ipersistent.ipersistent_iid);
    try printTuid(stdout, "IAttributes", base.ipersistent.iattributes_iid);
    try printTuid(stdout, "IAttributes2", base.ipersistent.iattributes2_iid);
}

fn printType(writer: anytype, comptime name: []const u8, comptime Type: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(Type), @alignOf(Type) });
}

fn printOffset(writer: anytype, comptime type_name: []const u8, comptime field_label: []const u8, comptime Type: type, comptime field_name: []const u8) !void {
    try writer.print("{s}.{s} offset {}\n", .{ type_name, field_label, @offsetOf(Type, field_name) });
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}

File: tests/abi/base_persistence_layout.cpp
#include "pluginterfaces/base/fvariant.h"
#include "pluginterfaces/base/icloneable.h"
#include "pluginterfaces/base/ipersistent.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Type), alignof (Steinberg::Type))
#define PRINT_OFFSET(Type, Field) std::printf (#Type "." #Field " offset %zu\n", offsetof (Steinberg::Type, Field))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	std::printf ("FVariant.kEmpty %u\n", Steinberg::FVariant::kEmpty);
	std::printf ("FVariant.kInteger %u\n", Steinberg::FVariant::kInteger);
	std::printf ("FVariant.kFloat %u\n", Steinberg::FVariant::kFloat);
	std::printf ("FVariant.kString8 %u\n", Steinberg::FVariant::kString8);
	std::printf ("FVariant.kObject %u\n", Steinberg::FVariant::kObject);
	std::printf ("FVariant.kOwner %u\n", Steinberg::FVariant::kOwner);
	std::printf ("FVariant.kString16 %u\n", Steinberg::FVariant::kString16);
	PRINT_TYPE (FVariant);
	PRINT_OFFSET (FVariant, type);
	PRINT_OFFSET (FVariant, intValue);
	PRINT_OFFSET (FVariant, floatValue);
	PRINT_OFFSET (FVariant, string8);
	PRINT_OFFSET (FVariant, string16);
	PRINT_OFFSET (FVariant, object);

	PRINT_TYPE (ICloneable);
	PRINT_TYPE (IPersistent);
	PRINT_TYPE (IAttributes);
	PRINT_TYPE (IAttributes2);
	print_iid ("ICloneable", Steinberg::ICloneable_iid);
	print_iid ("IPersistent", Steinberg::IPersistent_iid);
	print_iid ("IAttributes", Steinberg::IAttributes_iid);
	print_iid ("IAttributes2", Steinberg::IAttributes2_iid);
	return 0;
}
